### ISSUE TITLE
bug(`forge script`): buffer overrun while deserializing

### ISSUE THREAD

**Issue by lmy375** [foundry-rs/foundry#8465]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (af97b2c 2024-07-18T00:22:15.301937000Z)

### What command(s) is the bug in?

forge script

### Operating System

macOS (Apple Silicon)

### Describe the bug

When I'm trying deploying the contract with `forge script`, it crashes.

Script is like: 
```
address constant ZERO = address(0xc0b0);
address dummy = new MyContract(ZERO, ZERO);
```

Error message:
```
[⠒] Compiling...
No files changed, compilation skipped
Enter keystore password:
Script ran successfully.

## Setting up 1 EVM.
2024-07-18T03:08:39.237259Z ERROR forge_script::transaction: Failed to decode constructor arguments contract=Some("MyContract") signature=constructor(address,address) is_create2=false constructor_args=0000000000000000000000c0b0000000000000000000000000000000000000000000000000000000000000c0b0
Error: 
buffer overrun while deserializing

Context:
- buffer overrun while deserializing
```

It looks like the `constructor_args`, which should be 64 bytes but 45 byte here, is incorrectly sliced.

**Comment by zerosnacks** [comment]
Likely related: https://github.com/foundry-rs/foundry/issues/8491

**Comment by diyahir** [comment]
same issue

**Comment by zerosnacks** [comment]
@diyahir could you expand on the issue and provide a minimal example?

**Comment by yungwkndllc** [comment]
running `foundryup` fixed this for me

**Comment by zerosnacks** [comment]
Thanks for flagging @yungwkndllc 

Optimistically marking as `resolved`, @lmy375 / @diyahir please leave a comment / re-open with a minimal example if this is still an issue you are facing with the latest version of Foundry after running `foundryup`

**Comment by kris-gaudel** [comment]
Unfortunately running `foundryup` hasn't worked for me. The error is periodic in nature for me (i.e., I restart anvil and re-run my deploy script and it goes away). Are there any other fixes?

**Comment by mdulin2** [comment]
I'm running into the same issue. Seems to be somewhat random across builds that I've done.

**Comment by zerosnacks** [comment]
Hi @PatrickAlphaC, reopened the ticket

Would you have additional details or a minimal reproduction?

Thanks!

**Comment by fragtex-eth** [comment]
same issue

**Comment by zerosnacks** [comment]
Hi @fragtex-eth would you mind sharing more details? Preferably a minimal reproduction. Thanks!

**Comment by fragtex-eth** [comment]
Hi @zerosnacks, I fixed the issue by running forge clean and unfortunately can't reproduce it anymore. My script was similar to the one mentioned in this post, and I was using the latest stable release on Mac (with Apple Silicon)

**Comment by grandizzy** [comment]
Please provide a way to reproduce if the issue still persists after installing latest foundry and after a `forge clean`. Thank you!

**Comment by grandizzy** [comment]
> 
> On the latest stable release of foundry by running `./scripts/deploy_anvil.sh` from the `contracts/` directory of https://github.com/worldcoin/world-chain/tree/osiris/buffer-overrun-poc/contracts

@0xOsiris thank you for details, was able to reproduce with given steps, will check why this happens
- start anvil
`anvil --chain-id 31337 --block-time 2`
- run script
```
export PRIVATE_KEY=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
forge script scripts/DeployDevnet.s.sol:DeployDevnet --rpc-url http://127.0.0.1:8545 --chain-id $CHAIN_ID --broadcast
```

results in
```
== Logs ==
  Deploying: EntryPoint, PBHEntryPoint, PBHEntryPointImplV1, PBHSignatureAggregator, PBHSafe4337Module, WorldChainBlockRegistry
  PBHEntryPointImplV1 Deployed at:  0x5FbDB2315678afecb367f032d93F642f64180aa3
  PBHEntryPoint Deployed at:  0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
  PBHSignatureAggregator Deployed at:  0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0
2025-03-31T18:09:19.617098Z ERROR forge_script::transaction: Failed to decode constructor arguments contract=Some("ERC1967Proxy") signature=constructor(address,bytes) is_create2=false constructor_args=5b3d5ffd0000000000000000000000005fbdb2315678afecb367f032d93f642f64180aa30000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000010401a8afba0000000000000000000000005fbdb2315678afecb367f032d93f642f64180aa30000000000000000000000000000000071727de22e5e9d8baf0edac6f37da03200000000000000000000000000000000000000000000000000000000000000ff0000000000000000000000000000000000000000000000000000000001c9c38000000000000000000000000000000000000000000000000000000000000000c00000000000000000000000001804c8ab1f12e6bbf3894d4083f33e07309d1f38000000000000000000000000000000000000000000000000000000000000000100000000000000000000000070997970c51812dc3a010c7d01b50e0d17dc79c800000000000000000000000000000000000000000000000000000000
Error: ABI decoding failed: buffer overrun while deserializing
```

**Comment by 0xOsiris** [comment]
> > On the latest stable release of foundry by running `./scripts/deploy_anvil.sh` from the `contracts/` directory of https://github.com/worldcoin/world-chain/tree/osiris/buffer-overrun-poc/contracts
> 
> [@0xOsiris](https://github.com/0xOsiris) thank you for details, was able to reproduce with given steps, will check why this happens
> 
> * start anvil
>   `anvil --chain-id 31337 --block-time 2`
> * run script
> 
> ```
> export PRIVATE_KEY=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
> forge script scripts/DeployDevnet.s.sol:DeployDevnet --rpc-url http://127.0.0.1:8545 --chain-id $CHAIN_ID --broadcast
> ```
> 
> results in
> 
> ```
> == Logs ==
>   Deploying: EntryPoint, PBHEntryPoint, PBHEntryPointImplV1, PBHSignatureAggregator, PBHSafe4337Module, WorldChainBlockRegistry
>   PBHEntryPointImplV1 Deployed at:  0x5FbDB2315678afecb367f032d93F642f64180aa3
>   PBHEntryPoint Deployed at:  0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
>   PBHSignatureAggregator Deployed at:  0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0
> 2025-03-31T18:09:19.617098Z ERROR forge_script::transaction: Failed to decode constructor arguments contract=Some("ERC1967Proxy") signature=constructor(address,bytes) is_create2=false constructor_args=5b3d5ffd0000000000000000000000005fbdb2315678afecb367f032d93f642f64180aa30000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000010401a8afba0000000000000000000000005fbdb2315678afecb367f032d93f642f64180aa30000000000000000000000000000000071727de22e5e9d8baf0edac6f37da03200000000000000000000000000000000000000000000000000000000000000ff0000000000000000000000000000000000000000000000000000000001c9c38000000000000000000000000000000000000000000000000000000000000000c00000000000000000000000001804c8ab1f12e6bbf3894d4083f33e07309d1f38000000000000000000000000000000000000000000000000000000000000000100000000000000000000000070997970c51812dc3a010c7d01b50e0d17dc79c800000000000000000000000000000000000000000000000000000000
> Error: ABI decoding failed: buffer overrun while deserializing
> ```
> 
> (Pretty sure related to using the proxy contract rather than the PBHEntryPoint - https://github.com/worldcoin/world-chain/blob/bd1cd9c1cf79b76162b8551ae61df4180a2ef79e/contracts/src/PBHEntryPoint.sol#L28C5-L31) Most probably similar with [#9050](https://github.com/foundry-rs/foundry/pull/9050) where we had to use both deployed and creation code to identify proper artifact

Yes, it's certainly related to wrapping the `PBHEntryPoint` around the `ERC1967Proxy` and (maybe) grabbing the wrong artifact. I can confirm that when I change the deployment to set the proxy as `ERC1967Proxy` the error does not pop up.

**Comment by grandizzy** [comment]
yep, can confirm is similar as if you remove
```
bytecode_hash = "none"
cbor_metadata = false
```
from foundry.toml then it works

**Comment by grandizzy** [comment]
What we can do is to order artifacts and have project artifacts taking precedence on deps / libraries (like try to first match on project src, test, script and then on lib)
Right now we only order them by bytecode len but this can result in evaluating external contracts before project contracts

https://github.com/foundry-rs/foundry/blob/7a7ad4ea7282c30477d7fc0f39bf631acde7e7bd/crates/evm/traces/src/identifier/local.rs#L19-L26

@klkvr pls chime in when time, thank you!
